### PR TITLE
Static Site Builds

### PR DESCRIPTION
**Pull request by amiller-gh**
Okay! Here's take two at addressing #9 🙂🎉

I think this is just about ready to land – and is 1000% ready to be torn apart for review

I went ahead and implemented a slew of the architectural changes I mentioned in #99, making this one pretty meaty (sorry in advance...)

List of major changes includes:
  - Add `vapid build` cli command.
  - Break out core /lib/runners/Vapid.js file for config ingestion.
  - `lib/runners/VapidServer.js` extends core `Vapid` class for dev server.
  - `lib/runners/VapidBuilder.js` extends core `Vapid` class for static build.
  - Static builder generates static site under `/dist`.
  - Lots of Vapid specific path logic abstracted out to `lib/utils/paths.js` util file.
  - Added `/dist` to template .gitignore.
  - Added `build` script to template package.json
  - Pulled template rendering functionality to `lib/Renderer.js` util for sharing.
  - Abstracted some generic middleware business logic, path resolution, renderer helpers, and favicon discovery function to `lib/utils`.
  - Broke out Database functionality to its own standalone package folder.
  - Moved Template related files to its own standalone package folder.
  - De-coupled the Webpack config generator from the runner implementation.
  - Grouped utilities into their own standalone package folder.

The two more contentious changes wrapped up in this PR are:
 - Removed dependency on all of lodash.

    Depending on all of Lodash brought along a lot of unnecessary code bloat for just a few utility functions. I re-factored code to use well-supported native JavaScript methods where possible and added the few non-native utility functions directly to `/lib/utils`. This will reduce dependency size, and make the code more welcoming for those not familiar with the lodash utility functions.
    
 - Added ability to reference the general scope from within any section.
    
    The *one* feature addition that is honestly just in here because I desperately needed it for one of my projects and it accidentally got wrapped up in this commit 😅 Tests are added. 
    
    tl;dr – previously it was impossible to reference the `general` scope from within a child section. This can be really inconvenient for users. Consider my real-world example: I have a client with a site that links out to a 3rd party donation page. The link to this donation landing page may change over time, but will always be global across the site. Donation links being what they are, this link should be scattered throughout the site – both in the main nav bar, and on individual pages. Previously, I could not allow the client to simply set the URL once in the `general` scope and reference it in any section. Each section would need to have a separate donation URL. Ex:
    ```hbs 
    <nav><a href="{{donation_url}}>Donate</a></nav>
    {{#section subPage}} <a href="{{donation_url}}">Donate</a>{{/section}}
    ```
    It is now possible to reference the `general` scope (and *only* the general scope) within sections:
    ```hbs 
    <nav><a href="{{donation_url}}>Donate</a></nav>
    {{#section subPage}} <a href="{{general.donation_url}}">Donate</a>{{/section}}
    ```
    Like all other sections, the `general` data model will represent the union of all references across section definitions, even when only referenced inside a section:
    ```hbs 
    <nav><a href="{{donation_url}}>Donate</a></nav>
    {{#section subPage}} 
      <a href="{{general.donation_url}}">Donate</a>
      <a href="{{general.another_global_value}}">Donate</a>
    {{/section}}
    ```

There are still a few nice things I'd like to add:
 - Create a cleaner construct for `/form.js`
 - Better standardize the interface for "runners" (currently, the server uses `.start()` and `.stop()` while the builder uses `.build()`. Ideally, they have a unified interface – the server just doesn't resolve until it is explicitly stopped.
 - ~~Continue to simplify the `utility` package and more the code to more vanilla code (ex: `startsWith`, `endsWith` and `isEmpty` can likely all be replaced with native methods).~~ Edit: Done
 - ~~I have ideas about the template parser – this is for another PR though 🙂~~ Edit: see #158 

Anyway! That's a lot. Looking forward to feedback!

### CODE REVIEW

**Comment by amiller-gh** [comment]
Aaand forgot to run the linter before pushing.... will fix the style lint errors. 
Proposal: run lint on precommit?

**Comment by srobbin** [comment]
@amiller-gh Is this ready to go now? I saw the fix commits at the end of the week, so I wanted to hold off on a final review until you were ready.

**Comment by amiller-gh** [comment]
afaict this should be ready to go! Seems the test suite is missing a few tests for multiple sections and section ordering – I may come in and add those in a fast follow. Did a round of manual testing though and everything checks out.

I may be wrong but... it seems like custom section ordering may have been broken for some time? The Record default ordering seems to override the Section query order settings.

**Comment by amiller-gh** [comment]
Hmkay, spoke too soon. Found two more regressions added/modified tests for both so they won't fall through the cracks again 👍

**Comment by srobbin** [comment]
@amiller-gh Sorry for the slow review of this. I had noticed that the PR broke deployment, but didn't have a chance to sit down and figure out why until now. It looks like the `deploy` command is being passed an instance of `VapidServer` which no longer has `vapid.builder.tree`. I imagine we'll want to move deploy into a proper `VapidDeploy` runner, at least until we sort out the details of #157. I should have some time this weekend to convert deploy into a runner.

**Comment by srobbin** [comment]
@amiller-gh Let me know what you think of the VapidDeployer class—if it jives with the other runners. I still need to write some tests for it.

**Comment by amiller-gh** [comment]
Exciting! Glad you caught it – I don't have a Vapid deployment to test that flow on... Will take a look right now 👍

**Comment by srobbin** [comment]
Ah, right! I'll send you an invite for a free plan now.

File: lib/runners/VapidDeployer/index.js [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+      this.content = await this.getContent();
+
+      if (fs.existsSync(this.paths.uploads)) {
+        Utils.copyFiles(this.paths.uploads, path.resolve(this.tmpDir, 'uploads'));
+      }
+    }
+
+  return response.body.site.id;
+
+    return site;
```
Comment (amiller-gh) [inline lib/runners/VapidDeployer/index.js]: Double return!

**Comment by amiller-gh** [comment]
I also just pushed one last fix that I've had sitting locally the last week and a half. I botched a `_.each` => `for..of` conversion in the file uploader, so file uploads broke! We should really have a test for that – didn't get around to writing it as part of the fix...

Otherwise, I've been using this in production for the past two weeks with no other hiccups!

**Comment by amiller-gh** [comment]
Also, from my list of wants in the main PR text: 
> Better standardize the interface for "runners" (currently, the server uses `.start()` and `.stop()` while the builder uses `.build()`. Ideally, they have a unified interface – the server just doesn't resolve until it is explicitly stopped.

I think the different runner interfaces – including the `.build()` call you introduced here – could fit nicely into some kind of standard we can continue to flesh out in #157. Totally fine as is in the context of this PR, but would still love to see them fit into a well-defined pattern later as we keep iterating on a plugin-like system :)

File: lib/runners/VapidDeployer/index.js (hunk 1 above)
Comment (srobbin) [reply lib/runners/VapidDeployer/index.js]: I like to keep a spare return around. You never know when you'll need one! 😜

**Comment by srobbin** [comment]
I think it's going to take me a little longer to figure out how to test VapidDeployer, and I don't want to hold this PR up any longer, so I stuck in a placeholder for now.

Merging!

**Comment by amiller-gh** [comment]
🎉🎉🎉
